DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/Rules/AnonymizerTagRule.cs

```diff
@@ -0,0 +1,40 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Dicom;
+using EnsureThat;
+using Microsoft.Health.Dicom.Anonymizer.Core.Exceptions;
+using Microsoft.Health.Dicom.Anonymizer.Core.Model;
+using Microsoft.Health.Dicom.Anonymizer.Core.Processors;
+using Microsoft.Health.Dicom.Anonymizer.Core.Processors.Settings;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Health.Dicom.Anonymizer.Core.Rules
+{
+    public class AnonymizerTagRule : AnonymizerRule
+    {
+        public AnonymizerTagRule(DicomTag tag, string method, JObject ruleSetting, string description, IAnonymizerProcessorFactory processorFactory = null, IAnonymizerSettingsFactory settingsFactory = null)
+            : base(method, description, ruleSetting, processorFactory, settingsFactory)
+        {
+            EnsureArg.IsNotNull(tag, nameof(tag));
+
+            Tag = tag;
+        }
+
+        public DicomTag Tag { get; set; }
+
+        public override List<DicomItem> LocateDicomTag(DicomDataset dataset, ProcessContext context)
+        {
+            EnsureArg.IsNotNull(dataset, nameof(dataset));
+            EnsureArg.IsNotNull(context, nameof(context));
+
+            var locatedItems = new List<DicomItem>() { dataset.GetDicomItem<DicomItem>(Tag) };
+            return locatedItems.Where(x => x != null && !context.VisitedNodes.Contains(x.Tag.ToString())).ToList();
```

Comment:
> locatedItems.Where(x => x != null && !context.VisitedNodes.Contains(x.Tag.ToString())) [](http://example.com/codeflow?start=18&length=87)

just curious, is it possible that you run two times and you get the different results? For instance, you run rule I and a tag will be added into the visitedNodes, and then you run rule II and the same tag is selected in locatedItems, but since it is visited, it will not be processed. When the rules are loaded in reversed order, this rule II will process the item and rule I will ignore it. Is this a possible situation?

Reply:
Yes, every item can only be processed once and the order of rules will effect the results.